### DISCUSSION TITLE
Milvus durability and data integrity during insert/upsert.

### DISCUSSION THREAD

**Discussion by xiaobingxia-at** [milvus-io/milvus#37391] [Q&A and General discussion] [answered]
Hello Milvus Experts, let's say I am doing the MilvusClient.insert/upsert(....) with a lot of data, or I'm doing the collection.flush(). 
Either way, if my client crashed or the milvus cluster crashed, will my data on S3 or Milvus cluster be corrupted?
How does the Milvus to protect the data integrity in such "client crashing during insert" or "Milvus cluster crashing during insert" scenarios?

**Comment by xiaofan-luan** [comment, marked as answer]
no, because all the data is persistent in either kafka/pulsar or minio.

if write succeed all the data will be kept in kafka and during the recovery it is replayed

**Reply by xiaobingxia-at** [reply]
If the data persistence to kafka or s3, for one batch data, is ongoing, but not complete yet. At this time, the kafka or Milvus cluster is killed. Can we recover the data to the point before this batch of data persistence? (this batch of data will be completely discarded and cleaned by garbage collection)

**Reply by xiaofan-luan** [reply]
> If the data persistence to kafka or s3, for one batch data, is ongoing, but not complete yet. At this time, the kafka or Milvus cluster is killed. Can we recover the data to the point before this batch of data persistence? (this batch of data will be completely discarded and cleaned by garbage collection)

There is no atomicity guarantee right now about one batch of data. Under each shard it is one batch, but between multiple shard there is no guarantee
